lib/src/map/flutter_map_state.dart

```diff
@@ -0,0 +1,443 @@
+  // impossible (negative) value initially and only change it once Flutter
+  // provides real constraints.
+  static const kImpossibleSize = CustomPoint<double>(-1, -1);
+
+  final MapOptions options;
+
+  final LatLng center;
+  final double zoom;
+  final double rotation;
+
+  // Original size of the map where rotation isn't calculated
+  final CustomPoint<double> nonRotatedSize;
+
+  // Extended size of the map where rotation is calculated
+  final CustomPoint<double> size;
+
+  // Lazily calculated fields.
+  Bounds<double>? _pixelBounds;
+  LatLngBounds? _bounds;
+  CustomPoint<int>? _pixelOrigin;
+
+  MapSafeArea? _mapSafeAreaCache;
+
+  static FlutterMapState? maybeOf(BuildContext context) => context
+      .dependOnInheritedWidgetOfExactType<MapStateInheritedWidget>()
+      ?.mapState;
+
+  static FlutterMapState of(BuildContext context) =>
+      maybeOf(context) ??
+      (throw StateError(
+          '`FlutterMapState.of()` should not be called outside a `FlutterMap` and its descendants'));
+
+  /// Initializes FlutterMapState from the given [options] and with the
+  /// [nonRotatedSize] and [size] both set to [kImpossibleSize].
+  FlutterMapState.initialState(this.options)
+      : center = options.center,
+        zoom = options.zoom,
+        rotation = options.rotation,
+        nonRotatedSize = kImpossibleSize,
+        size = kImpossibleSize;
+
+  // Create an instance of FlutterMapState. The [pixelOrigin], [bounds], and
+  // [pixelBounds] may be set if they are known already. Otherwise if left
+  // null they will be calculated lazily when they are used.
+  FlutterMapState({
+    required this.options,
+    required this.center,
+    required this.zoom,
+    required this.rotation,
+    required this.nonRotatedSize,
+    required this.size,
+    Bounds<double>? pixelBounds,
+    LatLngBounds? bounds,
+    CustomPoint<int>? pixelOrigin,
+  })  : _pixelBounds = pixelBounds,
+        _bounds = bounds,
+        _pixelOrigin = pixelOrigin;
+
+  FlutterMapState withNonRotatedSize(CustomPoint<double> nonRotatedSize) {
+    if (nonRotatedSize == this.nonRotatedSize) return this;
+
+    return FlutterMapState(
+      options: options,
+      center: center,
+      zoom: zoom,
+      rotation: rotation,
+      nonRotatedSize: nonRotatedSize,
+      size: _calculateSize(rotation, nonRotatedSize),
+    );
+  }
+
+  FlutterMapState withRotation(double rotation) {
+    if (rotation == this.rotation) return this;
+
+    return FlutterMapState(
+      options: options,
+      center: center,
+      zoom: zoom,
+      rotation: rotation,
+      nonRotatedSize: nonRotatedSize,
+      size: _calculateSize(rotation, nonRotatedSize),
+    );
+  }
+
+  FlutterMapState withOptions(MapOptions options) => FlutterMapState(
+        options: options,
+        center: center,
+        zoom: zoom,
+        rotation: rotation,
+        nonRotatedSize: nonRotatedSize,
+        size: _calculateSize(rotation, nonRotatedSize),
+      );
```

Comment:
Unless I'm missing something, this should be part of `copyWith`.

Reply:
The withRotation method returns the current FlutterMapState if the rotation doesn't change. This avoids recalculating the size as well as retaining lazy calculated fields (pixelBounds/bounds/pixelOrigin).